File: Day_3/main.py
import time

def is_symbol(c):
    return not (c == '.' or c.isdigit())

def is_part_number(i, j, grid, num_rows, num_columns):
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (-1, 1), (1, -1), (1, 1)]
    for di, dj in directions:
        new_i, new_j = i + di, j + dj
        if 0 <= new_i < num_rows and 0 <= new_j < num_columns:
            if is_symbol(grid[new_i][new_j]):
                return True
    return False

def main():
    start_time = time.time()

    file_name = "data.txt"
    with open(file_name, 'r') as file:
        lines = file.readlines()

    num_rows = len(lines)
    num_columns = len(lines[0].strip())  # Ensure to strip newline characters
    grid = [list(line.strip()) for line in lines]  # Strip each line

    input_time = time.time()

    print((input_time - start_time) * 1000)

    current_num = ""
    total_sum = 0
    is_valid = False

    for i in range(num_rows):
        for j in range(num_columns):

            if grid[i][j].isdigit():
                current_num += grid[i][j]
                if not is_valid and is_part_number(i, j, grid, num_rows, num_columns):
                    is_valid = True
            else:
                if is_valid and current_num:
                    total_sum += int(current_num)
                current_num = ""
                is_valid = False

    print(f"sum -> {total_sum}")

    end_time = time.time()
    print(f"elapsed time -> {(end_time - start_time) * 1000 } ms")

if __name__ == "__main__":
    main()

File: Day_3/src/Main.kt
import java.io.File
import kotlin.math.max
import kotlin.math.min
import kotlin.system.measureTimeMillis

fun main() {
    val time = measureTimeMillis {

        val fileName = "data.txt"
        val lines = File(fileName).readLines()

        val numRows = lines.size
        val numColumns = lines[0].length
        val grid = Array(numRows) { CharArray(numColumns) }

        for ((i, value) in lines.withIndex()) {
            for (j in lines[i].indices) {
                grid[i][j] = lines[i][j]
            }
        }

        fun isPartNumber(i: Int, j: Int) : Boolean {
            fun isSymbol(c : Char) : Boolean {
                return !(c == '.' || c.isDigit())
            }

            val minRow = max(0, i - 1)
            val minCol = max(0, j - 1)
            val maxRow = min(numRows - 1, i + 1)
            val maxCol = min(numColumns - 1, j + 1)

            return isSymbol(grid[minRow][j]) ||
                    isSymbol(grid[maxRow][j]) ||
                    isSymbol(grid[i][minCol]) ||
                    isSymbol(grid[i][maxCol]) ||
                    isSymbol(grid[minRow][minCol]) ||
                    isSymbol(grid[minRow][maxCol]) ||
                    isSymbol(grid[maxRow][minCol]) ||
                    isSymbol(grid[maxRow][maxCol])
        }

        var currentNum = ""
        var totalSum = 0
        var isValid = false

        for (i in 0..(numRows-1)) {
            for (j in 0..(numColumns-1)) {

                if (grid[i][j].isDigit()) {
                    currentNum.plus(grid[i][j])
                    if(!isValid && isPartNumber(i,j)) {
                        isValid = true
                    }
                }
                else {
                    if(isValid && currentNum.isNotEmpty()) {
                        totalSum += currentNum.toInt()
                    }
                    currentNum = ""
                    isValid = false
                }
            }
        }

        println("sum -> $totalSum")
    }
    println("elapsed time -> $time ms")
}
